MaterialLib/SolidModels/Ehlers-impl.h

```diff
@@ -560,6 +547,107 @@ typename SolidEhlers<DisplacementDim>::KelvinVector predict_sigma(
     return sigma_D - pressure * Invariants::identity2;
 }
 
+template <typename ResidualVector, typename KelvinVector>
+std::tuple<KelvinVector, PlasticStrain<KelvinVector>, double>
+splitSolutionVector(ResidualVector const& solution)
+{
+    static auto const size = KelvinVector::SizeAtCompileTime;
+    return std::forward_as_tuple(
+        solution.template segment<size>(size * 0),
+        PlasticStrain<KelvinVector>{solution.template segment<size>(size * 1),
+                                    solution(size * 2), solution(size * 2 + 1)},
+        solution(size * 2 + 2));
+}
+
+/// Returns new solution, corresponding state, and the linear solver with the
+/// last decomposition. In case of failure nothing is returned.
+template <int JacobianResidualSize, int DisplacementDim>
+boost::optional<std::tuple<
+    ProcessLib::KelvinVectorType<DisplacementDim>,
+    MaterialStateVariables<DisplacementDim>,
+    Eigen::FullPivLU<Eigen::Matrix<double, JacobianResidualSize,
+                                   JacobianResidualSize, Eigen::RowMajor>>>>
+newton(double const dt, MaterialPropertiesV const& mp,
+       typename SolidEhlers<DisplacementDim>::KelvinVector const& eps_D,
+       double const eps_V,
+       NumLib::NewtonRaphsonSolverParameters const& nonlinear_solver_parameters,
+       MaterialStateVariables<DisplacementDim> state,
```

Comment:
Good catch! It was necessary before introduction of agglomerated solution vector, because the state was modified (+=) inplace. Now only plastic strain is returned, and state is const&.